src/emd/rt_bse.F

```diff
@@ -0,0 +1,1420 @@
+         CPASSERT(converged)
+         DO j = 1, rtbse_env%n_spin
+            CALL cp_cfm_to_cfm(rtbse_env%rho_new(j)%matrix, rtbse_env%rho(j)%matrix)
+         END DO
+         ! Print the updated field
+         CALL output_field(rtbse_env)
+         ! If needed, print out the density matrix in MO basis
+         CALL output_mos_contravariant(rtbse_env, rtbse_env%rho, rtbse_env%rho_section)
+         ! Also handles outputting to memory
+         CALL output_moments(rtbse_env, rtbse_env%rho)
+         ! Output restart files, so that the restart starts at the following time index
+         CALL output_restart(rtbse_env, rtbse_env%rho, i + 1)
+      END DO
+      ! ********************** End the time loop **********************
+
+      ! Carry out the FT
+      CALL output_moments_ft(rtbse_env)
+      CALL output_polarizability(rtbse_env)
+
+      ! Deallocate everything
+      CALL release_rtbse_env(rtbse_env)
+
+      CALL timestop(handle)
+   END SUBROUTINE run_propagation_bse
+
+! **************************************************************************************************
+!> \brief Calculates the initial values, based on restart/scf density, and other non-trivial values
+!> \param rtbse_env RT-BSE environment
+!> \param qs_env Quickstep environment (needed for reference to previous calculations)
+!> \author Stepan Marek (09.24)
+! **************************************************************************************************
+   SUBROUTINE initialize_rtbse_env(rtbse_env)
+      TYPE(rtbse_env_type), POINTER                :: rtbse_env
+      CHARACTER(len=*), PARAMETER                  :: routineN = "initialize_rtbse_env"
+      TYPE(post_scf_bandstructure_type), POINTER   :: bs_env
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER    :: moments_dbcsr_p
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER    :: matrix_s
+      REAL(kind=dp), DIMENSION(:), POINTER         :: custom_ref_point, &
+                                                      occupations
+      REAL(kind=dp), DIMENSION(3)                  :: rpoint
+      INTEGER                                      :: i, k, handle
+
+      CALL timeset(routineN, handle)
+
+      ! Get pointers to parameters from qs_env
+      CALL get_qs_env(rtbse_env%qs_env, bs_env=bs_env, matrix_s=matrix_s)
+
+      ! ****** START MOMENTS OPERATOR CALCULATION
+      ! Construct moments from dbcsr
+      NULLIFY (moments_dbcsr_p)
+      ALLOCATE (moments_dbcsr_p(3))
+      DO k = 1, 3
+         ! Make sure the pointer is empty
+         NULLIFY (moments_dbcsr_p(k)%matrix)
+         ! Allocate a new matrix that the pointer points to
+         ALLOCATE (moments_dbcsr_p(k)%matrix)
+         ! Create the matrix storage - matrix copies the structure of overlap matrix
+         ! Ideally, a proper matrix creation as below should be done, but does not work? - Use copy instead
+         ! CALL dbcsr_create(moments_dbcsr_p(i)%matrix, name="moments", template=matrix_s(1)%matrix)
```

Comment:
I think dbcsr_create just sets up the distribution, but no matrix layout. Ironically, dbcsr_copy also incorporates the call to dbcsr_create.